From db3cc159d335e40c4a62f821ae4fec2589771e2e Mon Sep 17 00:00:00 2001
From: electrikjesus <electrikjesus@gmail.com>
Date: Sat, 10 Jun 2023 13:40:58 -0400
Subject: [PATCH 2/3] Allow loading advanced power menu defaults

This change introduces a way to set and load defaults for showing the advanced power-menu

Overlays should target (frameworks/base/packages/SettingsProvider/res/values/defaults.xml):

<!-- default setting for Settings.System.POWERMENU_ADVANCED -->
    <bool name="def_powermenu_advanced">true</bool>
---
 core/java/android/provider/Settings.java             |  1 +
 core/proto/android/providers/settings/system.proto   |  3 ++-
 packages/SettingsProvider/res/values/defaults.xml    |  3 +++
 .../validators/SystemSettingsValidators.java         |  1 +
 .../providers/settings/SettingsProtoDumpUtil.java    |  3 +++
 .../android/providers/settings/SettingsProvider.java | 12 ++++++++++++
 .../com/android/server/power/ShutdownThread.java     |  2 +-
 7 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index c5638a5da6700..e93da165af6c7 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -6826,6 +6826,7 @@ public final class Settings {
             PUBLIC_SETTINGS.add(HAPTIC_FEEDBACK_ENABLED);
             PUBLIC_SETTINGS.add(SHOW_WEB_SUGGESTIONS);
             PUBLIC_SETTINGS.add(VIBRATE_WHEN_RINGING);
+            PUBLIC_SETTINGS.add(POWERMENU_ADVANCED);
         }
 
         /**
diff --git a/core/proto/android/providers/settings/system.proto b/core/proto/android/providers/settings/system.proto
index f8b5b233c6072..c0baed3133ed3 100644
--- a/core/proto/android/providers/settings/system.proto
+++ b/core/proto/android/providers/settings/system.proto
@@ -241,8 +241,9 @@ message SystemSettingsProto {
     optional Volume volume = 33;
 
     optional SettingProto when_to_make_wifi_calls = 34 [ (android.privacy).dest = DEST_AUTOMATIC ];
+    optional SettingProto powermenu_advanced = 35 [ (android.privacy).dest = DEST_AUTOMATIC ];
 
     // Please insert fields in alphabetical order and group them into messages
     // if possible (to avoid reaching the method limit).
-    // Next tag = 35;
+    // Next tag = 36;
 }
diff --git a/packages/SettingsProvider/res/values/defaults.xml b/packages/SettingsProvider/res/values/defaults.xml
index cf1f9bd9db128..f6d74eb6dd9cb 100644
--- a/packages/SettingsProvider/res/values/defaults.xml
+++ b/packages/SettingsProvider/res/values/defaults.xml
@@ -189,6 +189,9 @@
     <!--  default setting for Settings.System.END_BUTTON_BEHAVIOR : END_BUTTON_BEHAVIOR_SLEEP -->
     <integer name="def_end_button_behavior">0x2</integer>
 
+    <!-- default setting for Settings.System.POWERMENU_ADVANCED -->
+    <bool name="def_powermenu_advanced">false</bool>
+
     <!-- default setting for Settings.Global.DEFAULT_RESTRICT_BACKGROUND_DATA -->
     <bool name="def_restrict_background_data">false</bool>
 
diff --git a/packages/SettingsProvider/src/android/provider/settings/validators/SystemSettingsValidators.java b/packages/SettingsProvider/src/android/provider/settings/validators/SystemSettingsValidators.java
index 69a40d3894e48..6115fc3223b99 100644
--- a/packages/SettingsProvider/src/android/provider/settings/validators/SystemSettingsValidators.java
+++ b/packages/SettingsProvider/src/android/provider/settings/validators/SystemSettingsValidators.java
@@ -167,6 +167,7 @@ public class SystemSettingsValidators {
         VALIDATORS.put(System.MEDIA_BUTTON_RECEIVER, COMPONENT_NAME_VALIDATOR);
         VALIDATORS.put(System.HIDE_ROTATION_LOCK_TOGGLE_FOR_ACCESSIBILITY, BOOLEAN_VALIDATOR);
         VALIDATORS.put(System.VIBRATE_WHEN_RINGING, BOOLEAN_VALIDATOR);
+        VALIDATORS.put(System.POWERMENU_ADVANCED, BOOLEAN_VALIDATOR);
         VALIDATORS.put(System.DTMF_TONE_TYPE_WHEN_DIALING, BOOLEAN_VALIDATOR);
         VALIDATORS.put(System.HEARING_AID, BOOLEAN_VALIDATOR);
         VALIDATORS.put(System.TTY_MODE, new InclusiveIntegerRangeValidator(0, 3));
diff --git a/packages/SettingsProvider/src/com/android/providers/settings/SettingsProtoDumpUtil.java b/packages/SettingsProvider/src/com/android/providers/settings/SettingsProtoDumpUtil.java
index 4ac19386db0a6..59b9b050902f4 100644
--- a/packages/SettingsProvider/src/com/android/providers/settings/SettingsProtoDumpUtil.java
+++ b/packages/SettingsProvider/src/com/android/providers/settings/SettingsProtoDumpUtil.java
@@ -2741,6 +2741,9 @@ class SettingsProtoDumpUtil {
         dumpSetting(s, p,
                 Settings.System.END_BUTTON_BEHAVIOR,
                 SystemSettingsProto.END_BUTTON_BEHAVIOR);
+        dumpSetting(s, p,
+                Settings.System.POWERMENU_ADVANCED,
+                SystemSettingsProto.POWERMENU_ADVANCED);
         dumpSetting(s, p,
                 Settings.System.FONT_SCALE,
                 SystemSettingsProto.FONT_SCALE);
diff --git a/packages/SettingsProvider/src/com/android/providers/settings/SettingsProvider.java b/packages/SettingsProvider/src/com/android/providers/settings/SettingsProvider.java
index 53128848bbcb5..a902116da8d31 100644
--- a/packages/SettingsProvider/src/com/android/providers/settings/SettingsProvider.java
+++ b/packages/SettingsProvider/src/com/android/providers/settings/SettingsProvider.java
@@ -5222,6 +5222,18 @@ public class SettingsProvider extends ContentProvider {
                                 currentValue, null /* tag */, false /* makeDefault */,
                                 SettingsState.SYSTEM_PACKAGE_NAME);
                     }
+
+                    // Add default advanced power menu behavior
+                    final SettingsState systemSettings = getSystemSettingsLocked(userId);
+                    final Setting currentSetting = systemSettings.getSettingLocked(
+                            Settings.System.POWERMENU_ADVANCED);
+                    if (currentSetting.isNull()) {
+                        systemSettings.insertSettingOverrideableByRestoreLocked(
+                                Settings.System.POWERMENU_ADVANCED,
+                                getContext().getResources().getBoolean(
+                                        R.bool.def_powermenu_advanced) ? "1" : "0",
+                                null, true, SettingsState.SYSTEM_PACKAGE_NAME);
+                    }
                     currentVersion = 203;
                 }
 
diff --git a/services/core/java/com/android/server/power/ShutdownThread.java b/services/core/java/com/android/server/power/ShutdownThread.java
index f09d9496f2170..3767108abe904 100644
--- a/services/core/java/com/android/server/power/ShutdownThread.java
+++ b/services/core/java/com/android/server/power/ShutdownThread.java
@@ -259,7 +259,7 @@ public final class ShutdownThread extends Thread {
         mAdvancedReboot = false;
         mRebootSafeMode = false;
         mReason = reason;
-        mAdvancedReboot = true;
+        mAdvancedReboot = false;
         shutdownInner(context, confirm);
     }
 
-- 
2.34.1

